File: apis/src/main/java/com/beat/apis/booking/application/dto/GuestBookingRetrieveResponse.java
package com.beat.apis.booking.application.dto;

import java.time.LocalDateTime;

import com.beat.apis.performance.application.dto.BankNameType;
import com.beat.apis.schedule.application.dto.ScheduleNumberType;
import com.beat.global.support.jackson.CdnImageUrl;

public record GuestBookingRetrieveResponse(
	Long bookingId,
	Long scheduleId,
	Long performanceId,
	String performanceTitle,
	LocalDateTime performanceDate,
	String performanceVenue,
	int purchaseTicketCount,
	ScheduleNumberType scheduleNumber,
	String bookerName,
	String performanceContact,
	BankNameType bankName,
	String accountNumber,
	String accountHolder,
	int dueDate,
	BookingStatusType bookingStatus,
	LocalDateTime createdAt,
	@CdnImageUrl String posterImage,
	int totalPaymentAmount
) {
	public static GuestBookingRetrieveResponse of(
		Long bookingId,
		Long scheduleId,
		Long performanceId,
		String performanceTitle,
		LocalDateTime performanceDate,
		String performanceVenue,
		int purchaseTicketCount,
		ScheduleNumberType scheduleNumber,
		String bookerName,
		String performanceContact,
		BankNameType bankName,
		String accountNumber,
		String accountHolder,
		int dueDate,
		BookingStatusType bookingStatus,
		LocalDateTime createdAt,
		String posterImage,
		int totalPaymentAmount
	) {
		return new GuestBookingRetrieveResponse(
			bookingId,
			scheduleId,
			performanceId,
			performanceTitle,
			performanceDate,
			performanceVenue,
			purchaseTicketCount,
			scheduleNumber,
			bookerName,
			performanceContact,
			bankName,
			accountNumber,
			accountHolder,
			dueDate,
			bookingStatus,
			createdAt,
			posterImage,
			totalPaymentAmount
		);
	}
}

File: apis/src/main/kotlin/com/beat/apis/booking/api/response/GuestBookingRetrieveResponse.kt
package com.beat.apis.booking.api.response

import com.beat.apis.booking.api.type.BookingStatusType
import com.beat.apis.booking.application.result.BookingRetrieveResult
import com.beat.apis.performance.api.type.BankNameType
import com.beat.apis.schedule.api.type.ScheduleNumberType
import com.beat.global.support.jackson.CdnImageUrl
import java.time.LocalDateTime

@ConsistentCopyVisibility
data class GuestBookingRetrieveResponse private constructor(
    val bookingId: Long?,
    val scheduleId: Long?,
    val performanceId: Long?,
    val performanceTitle: String?,
    val performanceDate: LocalDateTime?,
    val performanceVenue: String?,
    val purchaseTicketCount: Int,
    val scheduleNumber: ScheduleNumberType?,
    val bookerName: String?,
    val performanceContact: String?,
    val bankName: BankNameType?,
    val accountNumber: String?,
    val accountHolder: String?,
    val dueDate: Int,
    val bookingStatus: BookingStatusType?,
    val createdAt: LocalDateTime?,
    @field:CdnImageUrl val posterImage: String?,
    val totalPaymentAmount: Int,
) {
    companion object {
        fun from(result: BookingRetrieveResult): GuestBookingRetrieveResponse = GuestBookingRetrieveResponse(
            bookingId = result.bookingId,
            scheduleId = result.scheduleId,
            performanceId = result.performanceId,
            performanceTitle = result.performanceTitle,
            performanceDate = result.performanceDate,
            performanceVenue = result.performanceVenue,
            purchaseTicketCount = result.purchaseTicketCount,
            scheduleNumber = result.scheduleNumber?.let(ScheduleNumberType::valueOf),
            bookerName = result.bookerName,
            performanceContact = result.performanceContact,
            bankName = result.bankName?.let(BankNameType::valueOf),
            accountNumber = result.accountNumber,
            accountHolder = result.accountHolder,
            dueDate = result.dueDate,
            bookingStatus = result.bookingStatus?.let(BookingStatusType::valueOf),
            createdAt = result.createdAt,
            posterImage = result.posterImage,
            totalPaymentAmount = result.totalPaymentAmount,
        )
    }
}
